homeassistant/components/surepetcare/__init__.py

```diff
@@ -42,113 +41,63 @@
+            entry.data[CONF_PASSWORD],
+            auth_token=entry.data[CONF_TOKEN] if CONF_TOKEN in entry.data else None,
             api_timeout=SURE_API_TIMEOUT,
             session=async_get_clientsession(hass),
         )
     except SurePetcareAuthenticationError:
         _LOGGER.error("Unable to connect to surepetcare.io: Wrong credentials!")
         return False
     except SurePetcareError as error:
-        _LOGGER.error("Unable to connect to surepetcare.io: Wrong %s!", error)
+        _LOGGER.error("Unable to connect to surepetcare.io: %s", error)
         return False
 
-    spc = SurePetcareAPI(hass, surepy)
+    spc = SurePetcareAPI(hass, entry, surepy)
     hass.data[DOMAIN][SPC] = spc
 
-    await spc.async_update()
-
-    async_track_time_interval(hass, spc.async_update, SCAN_INTERVAL)
-
-    # load platforms
-    for platform in PLATFORMS:
-        hass.async_create_task(
-            hass.helpers.discovery.async_load_platform(platform, DOMAIN, {}, config)
-        )
-
-    async def handle_set_lock_state(call):
-        """Call when setting the lock state."""
-        await spc.set_lock_state(call.data[ATTR_FLAP_ID], call.data[ATTR_LOCK_STATE])
-        await spc.async_update()
-
-    lock_state_service_schema = vol.Schema(
-        {
-            vol.Required(ATTR_FLAP_ID): vol.All(
-                cv.positive_int, vol.In(spc.states.keys())
-            ),
-            vol.Required(ATTR_LOCK_STATE): vol.All(
-                cv.string,
-                vol.Lower,
-                vol.In(
-                    [
-                        # https://github.com/PyCQA/pylint/issues/2062
-                        # pylint: disable=no-member
-                        LockState.UNLOCKED.name.lower(),
-                        LockState.LOCKED_IN.name.lower(),
-                        LockState.LOCKED_OUT.name.lower(),
-                        LockState.LOCKED_ALL.name.lower(),
-                    ]
-                ),
-            ),
-        }
-    )
-
-    hass.services.async_register(
-        DOMAIN,
-        SERVICE_SET_LOCK_STATE,
-        handle_set_lock_state,
-        schema=lock_state_service_schema,
-    )
-
-    return True
+    return await spc.async_setup()
 
 
 class SurePetcareAPI:
     """Define a generic Sure Petcare object."""
 
-    def __init__(self, hass: HomeAssistant, surepy: Surepy) -> None:
+    def __init__(
+        self, hass: HomeAssistant, config_entry: ConfigEntry, surepy: Surepy
+    ) -> None:
         """Initialize the Sure Petcare object."""
+
         self.hass = hass
+        self.config_entry = config_entry
         self.surepy = surepy
-        self.states = {}
+        self.states: dict[int, Any] = {}
 
     async def async_update(self, _: Any = None) -> None:
         """Get the latest data from Sure Petcare."""
 
         try:
-            self.states = await self.surepy.get_entities()
+            self.states = await self.surepy.get_entities(refresh=True)
```

Comment:
I honestly think maybe implementing a data coordinator before the config flow might be beneficial for the integration overall.

Reply:
okay 👍